File: backend/src/application/agent/ProductSnapshot.ts
import { ProductListing } from "../../domain/entities/ProductListing.js";

export type ProductSnapshot = {
  id: string;
  shopifyProductId: string | null | undefined;
  title: string;
  handle: string | null | undefined;
  vendor: string | null | undefined;
  productType: string | null | undefined;
  description: string;
  tags: string[];
  price: number | null | undefined;
  currency: string;
  inventoryQuantity: number | null | undefined;
  rawData: Record<string, unknown>;
};

export class ProductSnapshotTool {
  execute(product: ProductListing): ProductSnapshot {
    return {
      id: product.id,
      shopifyProductId: product.shopifyProductId,
      title: product.title.trim(),
      handle: product.handle,
      vendor: product.vendor,
      productType: product.productType,
      description: product.description?.trim() ?? "",
      tags: product.tags.map((tag) => tag.trim()).filter(Boolean),
      price: product.price,
      currency: product.currency,
      inventoryQuantity: product.inventoryQuantity,
      rawData: product.rawData
    };
  }
}

File: app/services/agent/product_snapshot.rb
module Agent
  ProductSnapshot = Struct.new(
    :id, :shopify_product_id, :title, :handle, :vendor, :product_type,
    :description, :tags, :price, :currency, :inventory_quantity, :raw_data,
    keyword_init: true
  )

  class ProductSnapshotTool
    def execute(listing)
      ProductSnapshot.new(
        id: listing.id,
        shopify_product_id: listing.shopify_product_id,
        title: listing.title.to_s.strip,
        handle: listing.handle,
        vendor: listing.vendor,
        product_type: listing.product_type,
        description: listing.description.to_s.strip,
        tags: Array(listing.tags).map(&:strip).reject(&:blank?),
        price: listing.price,
        currency: listing.currency,
        inventory_quantity: listing.inventory_quantity,
        raw_data: listing.raw_data || {}
      )
    end
  end
end
